Source/Client.swift

```diff
@@ -139,7 +144,12 @@ extension Client {
     private func send(_ request: Request, middleware: [Middleware]) throws -> Response {
         var request = request
         addHeaders(to: &request)
-        return try middleware.chain(to: self).respond(to: request)
+        let response = try middleware.chain(to: self).respond(to: request)
+        if 400..<600 ~= response.status.statusCode {
```

Comment:
429 Too Many Requests does, for example. I think all <400 should probably close it, just to be safe. It's much harder to recover the it gets closed (code-wise) than to assume that when you get an error 4xx or 5xx, you'll want to "start over" anyway. We can change it later, it's transparent to users (it'll only result in a new connection being open, but again I think that's the proper way for error responses when servers close the connections quite often anyway).

If you find any literature on this, I'd love to read more on it, but this fixed the problems I was seeing and makes sense to me.